peekingduck/training/src/data/dataset.py

```diff
@@ -0,0 +1,224 @@
+
+        self.image_path = df[cfg.dataset.image_path_col_name].values
+        self.targets = df[cfg.dataset.target_col_id].values if stage != "test" else None
+
+    def __len__(self) -> int:
+        """Return the length of the dataset."""
+        return len(self.df.index)
+
+    def __getitem__(self, index: int) -> Union[Tuple, Any]:
+        """Generate one batch of data"""
+        image_path: str = self.image_path[index]
+        image: Tensor = cv2.imread(image_path)
+        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+        image = self.apply_image_transforms(image)
+
+        # Get target for all modes except for test dataset.
+        # If test, replace target with dummy ones as placeholder.
+        target = self.targets[index] if self.stage != "test" else torch.ones(1)
+        # target = self.apply_target_transforms(target)
+
+        # TODO: consider stage to be private since it is only used internally.
+        if self.stage in ["train", "valid", "debug"]:
+            return image, target
+        elif self.stage == "test":
+            return image
+        else:
+            raise ValueError(f"Invalid stage {self.stage}.")
+
+    def apply_image_transforms(self, image: torch.Tensor) -> Tensor:
+        """Apply transforms to the image."""
+        if self.transforms and isinstance(self.transforms, A.Compose):
+            image = self.transforms(image=image)["image"]
+        elif self.transforms and isinstance(self.transforms, T.Compose):
+            image = self.transforms(image)
+        else:
+            image = torch.from_numpy(image).permute(2, 0, 1)  # convert HWC to CHW
+        return image
+
+    # pylint: disable=no-self-use # not yet!
+    def apply_target_transforms(
+        self, target: torch.Tensor, dtype: torch.dtype = torch.long
+    ) -> torch.Tensor:
+        """Apply transforms to the target.
+        Note:
+            This is useful for tasks such as segmentation object detection where
+            targets are in the form of bounding boxes, segmentation masks etc.
+        """
+        return torch.tensor(target, dtype=dtype)
+
+
+class TFImageClassificationDataset(tf.keras.utils.Sequence):
+    """Template for Image Classification Dataset."""
+
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        stage: str = "train",
+        batch_size: int = 1,
+        num_classes: int = 2,
+        target_size: Union[list, tuple] = (32, 32),
+        shuffle: bool = False,
+        transforms: TransformTypes = None,
+        num_channels: int = 3,
+        x_col: str = "",
+        y_col: str = "",
+        **kwargs: Dict[str, Any],
+    ) -> None:
+        """"""
+
+        self.df = df
+        self.stage = stage
+        self.transforms = transforms
+        self.batch_size = batch_size
+        self.num_classes = num_classes
+        self.dim = target_size
+        self.num_channels = num_channels
+        self.shuffle = shuffle
+
+        self.list_IDs = df[x_col].values
```

Comment:
Consider following the snake case convention (which applies to the rest of the repo).